api/client/client_test.go

```diff
@@ -0,0 +1,487 @@
+	_, clean := server(t)
+	defer clean()
+	client := client(t)
+
+	storeID, _ := client.NewStore()
+	err := client.RegisterSchema(storeID, modelName, schema)
+	checkErr(t, err)
+	err = client.Start(storeID)
+	checkErr(t, err)
+
+	person := createPerson()
+
+	err = client.ModelCreate(storeID, modelName, person)
+	checkErr(t, err)
+
+	exists, err := client.ModelHas(storeID, modelName, person.ID)
+	if err != nil {
+		t.Fatalf("failed to check model has: %v", err)
+	}
+	if !exists {
+		t.Fatal("model should exist but it doesn't")
+	}
+}
+
+func TestModelFind(t *testing.T) {
+	_, clean := server(t)
+	defer clean()
+	client := client(t)
+
+	storeID, _ := client.NewStore()
+	err := client.RegisterSchema(storeID, modelName, schema)
+	checkErr(t, err)
+	err = client.Start(storeID)
+	checkErr(t, err)
+
+	person := createPerson()
+
+	err = client.ModelCreate(storeID, modelName, person)
+	checkErr(t, err)
+
+	q := es.JSONQuery{
+		Ands: []es.JSONCriterion{
+			es.JSONCriterion{
+				FieldPath: "lastName",
+				Operation: es.Eq,
+				Value: es.JSONValue{
+					String: &person.LastName,
+				},
+			},
+		},
+	}
+
+	jsonResults, err := client.ModelFind(storeID, modelName, q)
+	if err != nil {
+		t.Fatalf("failed to find: %v", err)
+	}
+	results := make([]*Person, len(jsonResults))
+	for i, jsonResult := range jsonResults {
+		person := &Person{}
+		if err := json.Unmarshal(jsonResult, person); err != nil {
+			t.Fatalf("failed to unmarshal json result: %v", err)
+		}
+		results[i] = person
+	}
+	if len(results) != 1 {
+		t.Fatalf("expected 1 result, but got %v", len(results))
+	}
+	if !reflect.DeepEqual(results[0], person) {
+		t.Fatal("model found by query does't equal the original")
+	}
+}
+
+func TestModelFindByID(t *testing.T) {
+	_, clean := server(t)
+	defer clean()
+	client := client(t)
+
+	storeID, _ := client.NewStore()
+	err := client.RegisterSchema(storeID, modelName, schema)
+	checkErr(t, err)
+	err = client.Start(storeID)
+	checkErr(t, err)
+
+	person := createPerson()
+
+	err = client.ModelCreate(storeID, modelName, person)
+	checkErr(t, err)
+
+	newPerson := &Person{}
+	err = client.ModelFindByID(storeID, modelName, person.ID, newPerson)
+	if err != nil {
+		t.Fatalf("failed to find model by id: %v", err)
+	}
+	if !reflect.DeepEqual(newPerson, person) {
+		t.Fatal("model found by id does't equal the original")
+	}
+}
+
+func TestReadTransaction(t *testing.T) {
+	_, clean := server(t)
+	defer clean()
+	client := client(t)
+
+	storeID, err := client.NewStore()
+	checkErr(t, err)
+	err = client.RegisterSchema(storeID, modelName, schema)
+	checkErr(t, err)
+	err = client.Start(storeID)
+	checkErr(t, err)
+	person := createPerson()
+	err = client.ModelCreate(storeID, modelName, person)
+	checkErr(t, err)
+
+	txn, err := client.ReadTransaction(storeID, modelName)
+	if err != nil {
+		t.Fatalf("failed to create read txn: %v", err)
+	}
+
+	defer func() {
+		err = txn.End()
+		if err != nil {
+			t.Fatalf("failed to end txn: %v", err)
+		}
+	}()
```

Comment:
You _must_ always end a transaction, even if there is an error during a transaction. This seems reasonable because we don't want to assume the transaction should be ended automatically if there is an error during the transaction.

Reply:
That's a clean close. I'm wonder if we can think about good practices of using these things, since it may feel a bit verbose that kind of defers. As a client, I'd generally ignore if ending the readtx had an error since sounds quite safe.

Also, we could make `txn.End()` be a return of `ReadTransaction` to insist on the fact that this is a task that should be done explicitly. (since maybe someone doesn't notice there is an `End()` method). Just playing around with an idea, not sure.

Reply:
Yea, I think once we get a feel for using the transactions, we can make tweaks, but I think you're idea of returning an `end` function from `txn.Start` could be cool. I'll try that out.

Reply:
Ok, `txn.Start` now returns the end function. I like it, check it out.

Reply:
Nice :+1: